### PR TITLE
Enumerable Methods

### PR DESCRIPTION
**Pull request by marcode95**
## Enumerable Methods

-created custom enumerable methods using basic ruby concepts
-made use of yield keyword for accessing blocks inside our methods
-ensured functionality of my_map for both blocks and procs
-ensured functionality of methods for both arrays and hashes

### CODE REVIEW

File: enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    if block_given?
+      new_arr = []
+      arr.my_each do |element|
+        next unless yield(element)
+
+        new_arr.push(element)
+      end
+      puts new_arr.length
+    else
+      puts arr.length
```
Comment (kitpao) [inline enumerable_methods.rb]: - [x] Please take out the `puts` command, since it only prints the result but doesn't return any value for later use

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @marcode95 and @Abhigyan001 

⭐ Great job, your repo looks professional (readme, commits and pull request) and Git Flow and linter setup is correct :)

There are some failing tests in your methods, kindly check them out and fix them 🔽 

Here is a description of the test cases that are not passing
### Methods:
#### In general
- [x] Please notice that the original methods take into account what happens when a block is not given (They return an enumerator object). This applies to my_each, my_each_with_index, my_select, my_map **Hint:** you can check [this documentation for more](https://apidock.com/ruby/Object/to_enum). 
- If you have more doubts about how the original methods work, you can check here 👉 https://ruby-doc.org/core-2.6.4/Enumerable.html
#### #my_each and #my_each_with_index
- [x] Please check and correct the difference between original and your methods:
  ```
  p [1,2,3].each{ |x| x }               # [1,2,3]
  p [1,2,3].my_each{ |x| x }             # nil
  ```
  Notice that the original returns the original array and yours is returning nil. **Hint:** Don't forget to return something at the end
#### #my_all?,  #my_any?, and #my_none?
- [x] Please check and correct the difference between original and your method:
  ```
  p [1,2,3].all?               # true
  p [1,2,3].my_all?            # LocalJumpError
  ```
  Notice that the original returns True and yours is returning LocalJumpError. **Hint:** These methods should work even if a block is not given, also observe the behavior of the case `p [false, true, true].all?`

- [x] Please notice that your methods should be able to take Classes, Regexp and Patterns as parameters, **Hint:** Investigate the "===" operator, it can be very useful for this
```
p [1,2,3].all?(Integer)              # true
p [1,2,3].my_all?(Integer)           # `my_all?': wrong number of arguments (given 1, expected 0) (ArgumentError)
```
#### #my_count
- [x] This method should be able to receive parameters and count the items that are the same, check the example:
```
p [1,1,2,2,2,2].count(2)        # 4
p [1,1,2,2,2,2].my_count(2)     # `my_count': wrong number of arguments (given 1, expected 0) (ArgumentError)
```
#### #my_inject
- [x] If there is no block or argument, your method should return a local jump error ([please check how to raise errors here](https://stackoverflow.com/questions/4325507/how-to-force-a-localjumperror-on-return-in-ruby))
- [x] Your method should be able to manage symbols for operations (:+, :*, etc)
- [x] Your method should also be able to multiply elements, not only add them `[2,3].my_inject(20, :*)` should return 120 (20 * 2 * 3)
- You can check more examples [here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject)
#### #multiply_els
  - [x] `multiply_els [2, 4, 5]` should return 40, your method is undefined for an array **Hint:** take that method out of the module and pass the arguments manually

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me in slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate if can leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😛 

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

File: enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    return enum_for(:my_each) unless block_given?
+
+    arr = *self
+    i = 0
+    while i < arr.length
+      yield(arr[i])
+      i += 1
+    end
+    arr
+  end
```
Comment (BrittanyBlake) [inline enumerable_methods.rb]: - [x] When `my_each` is called on a range and a block is given, it should return the range itself after calling the given block once for each element in the range. 

```ruby
block = proc { |num| num < 4 }
range = Range.new(5, 50) 
range.my_each(&block) === range.each(&block) => false
```

_Hint: Yours is currently returning all of the elements in the range and not the range._

File: enumerable_methods.rb [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    return enum_for(:my_each_with_index) unless block_given?
+
+    arr = *self
+    i = 0
+    while i < arr.length
+      yield(arr[i], i)
+      i += 1
+    end
+    arr
+  end
```
Comment (BrittanyBlake) [inline enumerable_methods.rb]: - [ ] `my_each_with_index` has the same problem as `my_each`. 
```ruby
block = proc { |num| num < 4 }
range = Range.new(5, 50) 
range.my_each_with_index(&block) === range.each_with_index(&block) => false
```

File: enumerable_methods.rb [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        else
+          if element == false || element.nil?
+            condition = false
+            break
+          end
+        end
+      end
+    end
+    condition
+  end
```
Comment (BrittanyBlake) [inline enumerable_methods.rb]: - [x] When a pattern other than a Regex or a Class is given, `my_all?` should return false if any of the members of the collection does not match the pattern. 
```ruby
[5, 3, 3, 2, 4, 8, 3, 6, 5, 0, 1, 5].my_all?(3) === [5, 3, 3, 2, 4, 8, 3, 6, 5, 0, 1, 5].all?(3) => false
```

File: enumerable_methods.rb [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        else
+          if element || element.nil?
+            condition = true
+            break
+          end
+        end
+      end
+    end
+    condition
+  end
```
Comment (BrittanyBlake) [inline enumerable_methods.rb]: - [x] When no block or argument is given, `my_any? `should return false if at least one of the members of the collection is not true
```ruby
[nil, false].any? === [nil, false].my_any?  => false
```
****

- [x]  When a pattern other than a Regex or a Class is given, it should return false if none of the members of the collection matches the pattern.
````ruby
words = %w[dog door rod blade]
words.my_any?('cat') === words.any?('cat') => false

File: enumerable_methods.rb [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+  # my_none?
+
+  def my_none?(arg = nil)
+    !my_any?(arg)
+  end
```
Comment (BrittanyBlake) [inline enumerable_methods.rb]: - [x] When no block or argument is given, `my_none` should only return true if none of the members of the collection is true
```ruby
 [nil, false].none? === [nil, false].my_none? => false
```
****

- [x] When a pattern other than a Regex or a Class is given, it should return true if none of the members of the collection matches the pattern.
```ruby
words = %w[dog door rod blade]
words.my_none?('cat') === words.none?('cat') => false
```

File: enumerable_methods.rb [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    elsif !block_given? && !arg2
+      result = 0
+      result = 1 if arg1 == (:* || '*')
+      my_each do |element|
+        result = result.send arg1, element
+      end
+    end
+    result
+  end
+end
```
Comment (BrittanyBlake) [inline enumerable_methods.rb]: - [x] `my_inject` should raise a `localJumpError` when no block or argument is given. 
    - Yours crrently retrieves a `typeError`
```ruby
array = [1,2,3,4,5]
array.my_inject
```
screenshot of the `my_inject` error message 👇 
Screen Shot 2020-09-22 at 8 08 31 PM

```ruby
array = [1,2,3,4,5]
array.inject
```
screenshot of the `inject` error message 👇 
Screen Shot 2020-09-22 at 8 10 08 PM
****

- [ ] `my_inject` should search for the longest word in an array of strings.
 ```ruby
search =  proc { |memo, word| memo.length > word.length ? memo : word } 
words = %w[dog door rod blade]
words.my_inject(&search) 
```
- When trying to test this, your code throws an error 👇 
Screen Shot 2020-09-22 at 8 16 37 PM
results from the real `inject` method 👇 
Screen Shot 2020-09-22 at 8 16 56 PM
****

- [x] When `my_inject` is called on a range and given a block without an initial value, it should combine all the elements of enum by applying a binary operation.
```ruby
 range = Range.new(5, 50)
range.my_inject { |prod, n| prod * n } === range.inject { |prod, n| prod * n } => false
```
_Hint: Yours returns 0_

**Review by BrittanyBlake — CHANGES_REQUESTED** [review]
### Status: Changes Required  🏗️ 🔨 
****
### Hello @marcode95 and @Abhigyan001   👋 
You are doing a great job with this challenging project! 👍 You are almost done! 💪 Great job applying the feedback you have received from the previous TSEs. 🎖️🌟 However, there are a few issues in the following methods that need to be fixed. 👇 Please take a look at my comments below for detailed information about the errors.  
****

- `my_each`
- `my_each_with_index`
- `my_all?`
- `my_any?`
- `my_none?`
- `my_inject`

****
### Kindly make the required changes and resubmit for a new code review once you are ready 🙏


Good Luck and Happy Coding 🍀 👩🏾‍💻

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Brittany Blake)._

File: enumerable_methods.rb [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+          end
+        end
+      end
+    end
+    condition
+  end
+
+  # my_any?
+
+  def my_any?(arg = nil)
```
Comment (tirthajyoti-ghosh) [inline enumerable_methods.rb:85]: - When no block or argument is given, **returns true** if at least one of the collections is not false or nil but your method returned `false`.

- When a pattern other than Regex or a Class is given, **returns true** if any of the collection matches the pattern but your method returned `false`.

File: enumerable_methods.rb [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+      elsif arg.nil? && !block_given?
+        break if element == false
+      end
+    end
+    condition
+  end
+
+  # my_none?
+
+  def my_none?(arg = nil)
```
Comment (tirthajyoti-ghosh) [inline enumerable_methods.rb:128]: - It should **return false** if the block ever returns true for all elements but your method returned `true`.

- It should **return false** if the block ever returns true for all elements::range but your method returned `true`.

- When no block or argument is given **returns false** only if one of the collection members is true but your method returned `true`.

File: enumerable_methods.rb [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+      my_each { |i| new_arr.push my_proc.call(i) }
+    else
+      my_each { |i| new_arr.push(yield(i)) }
+    end
+    new_arr
+  end
+
+  # my_inject
+
+  def my_inject(arg1 = 0, arg2 = nil)
```
Comment (tirthajyoti-ghosh) [inline enumerable_methods.rb]: - When a block is given with an initial value, it combines all elements of enum by applying a binary operation, specified by a block.

  image

- When a symbol is specified **without an initial value** combines each element of the collection by applying the symbol as a named method. Your method throws an error - `NoMethodError: undefined method zero?' for :+:Symbol`.

- When a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named **method:range**. Your method throws an error - `NoMethodError: undefined method zero?' for :+:Symbol`.

File: enumerable_methods.rb [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+  # my_injectt
+
+  def my_injectt
+    result = self[0]
+
+    my_each do |item|
+      result = yield(result, item) if block_given?
+    end
+    result
+  end
```
Comment (tirthajyoti-ghosh) [inline enumerable_methods.rb]: I think this method can be removed as it's not getting used anywhere.

File: enumerable_methods.rb [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+end
+
+def multiply_els(arg1 = nil)
```
Comment (tirthajyoti-ghosh) [inline enumerable_methods.rb:256]: It accepts an array as an argument and multiplies all the elements of the array together using #my_inject. Your method throws an error - `NoMethodError: undefined method zero?' for :+:Symbol`. 

This error is related to the `#my_inject`. So fixing that should fix this as well.

**Review by tirthajyoti-ghosh — CHANGES_REQUESTED** [review]
# Changes required 🛑 

## Impressive progress so far 💪! Thanks for making all the changes 👏.

### However, we need to work on some changes before we can merge.

_I have added some comments below. Be sure to check them._

Your project looks absolutely fantastic. Just make these changes and apply for another code review.

Happy coding 👨‍💻 

*Need help? Shoot me a message on Slack - [@Ghosh](https://app.slack.com/client/T47CT8XPG/DRDM6LAQN/user_profile/URB1MHXK6)*
<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Abhigyan Mahanta & Marco Erhardt** 
Great job following the improvements from the previous review :muscle:  :muscle: 
Your project meets most of the requirements :clap: :clap: :clap:  
We still need to fix some details:

- [x] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
      `['cat', 'dog','car'].any?('cat') == ['cat', 'dog','car'].my_any?('cat')` 
- [x] #my_inject raises a 'LocalJumpError' when no block or argument is given
    ```
    begin
      [1,2,3,4].my_inject
    rescue LocalJumpError
      true
    end
    ```
     
- [x] #my_inject does not mutate the original array
    ```
    array = [1,2,3,4]
    array_clone = array.clone
    array.my_inject{|total, num| total+num}
    array == array_clone
    ```
     
- [x] #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block
      `[1,2,3].inject(&proc{|total, num| total*num}) == [1,2,3].my_inject(&proc{|total, num| total*num})` 
- [x] #my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a block
      `(1..3).inject(4) { |prod, n| prod * n } == (1..3).my_inject(4) { |prod, n| prod * n }`

Please request another review once you are ready! :+1: 

Happy coding :relaxed:

Oscar Andrés Russi Porras

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Abhigyan Mahanta & Marco Erhardt** 
Nice job with the improvement :muscle:  :muscle: 
Your Enumerable methods passes all the automatic tests :clap: :clap: :clap:  
Just some minor details to fix:

- [x] Make sure rubocop doesn't mark any offenses in your code
- [x] Add a screenshot in your README file

Please request another review once you are ready! :+1: 

Happy coding :relaxed:

Oscar Andrés Russi Porras

**Review by elmejdki — APPROVED** [review]
# status: approved :heavy_check_mark: 
## Well done guys, thank you for making the required changes.

now you can merge to development your project is approved.

## Happy Coding :tophat:

**Pull request merged**